### DISCUSSION TITLE
Freeform / Stripe Webhook showing 500 and 404 errors

### DISCUSSION THREAD

**Discussion by ndsurgenor** [solspace/craft-freeform#1619] [Questions & Answers] [answered]
### What happened?

Using a form in Craft which is integrated with Stripe to process one-off payments. Data is being submitted OK and received in form submissions/email notifications but the webhooks for development and live site as displaying errors for the invoice.payment_succeeded and customer.subscription.created events

**DEV WEBHOOK**
<img width="1146" alt="Screenshot 2024-10-11 at 15 27 45" src="https://github.com/user-attachments/assets/5df341eb-38fe-42e6-9cd6-09a7e26b134e">

Connection denied (500 error shown using stripe CLI)

LIVE WEBHOOK
<img width="1151" alt="Screenshot 2024-10-11 at 15 28 58" src="https://github.com/user-attachments/assets/48eae401-48af-4918-9abb-453b27679b39">

404 Error

I've checked and double-checked the variables and url endpoints and they are all correct. As I said, the payment is being accepted when using test cards on the dev and through testing on live but these events are showing errors and, as stripe keeps retrying, it's eventually causing the webhook to disable.

I've chatted plenty with ChatGPT and its suggestion at this point is to start making changes to the vendor code which I'm not going to do!

### How can we reproduce this?

1. Send up Form in Craft
2. Integrate with Stripe
3. Submit form then inspect webhooks on Stripe

### Freeform Edition

Pro

### Freeform Version

5.1.12

### Craft Version

Pro 4.8.6

### When did this issue start?

Unsure

**Comment by gustavs-gutmanis** [comment]
Hi @ndsurgenor,

Sorry to hear you're experiencing issues.

Could you share a screenshot of your Stripe configuration inside the `Settings > Payments` page for both Live and Dev sites? (With sensitive data blurred out, of course)

From reading the ticket, first thing that comes to mind is a misconfiguration issue.

### For `dev`:
When you want to test events on a `dev` environment, you're most likely working on a local hosted server on your computer, and that address is not accessible from outside. To test events locally you would need to use the [Stripe CLI](https://docs.stripe.com/stripe-cli) and [forward webhook events](https://docs.stripe.com/webhooks#test-webhook) to your dev url.

To listen to successful payments, you would do:

```
$ stripe listen --forward-to https://your-local-url.test/freeform/payments/stripe/webhook --skip-verify --events=payment_intent.succeeded
```

Then, when making successful payments locally in your `dev` environment - you would receive proper webhook triggers.

### For `prod`:
To have webhooks working properly in `prod`, you need to make sure you create a new Webhook in your Stripe Dashboard. 

Specify the URL that is displayed in your Stripe integration page for **"Webhook URL"**,

![image](https://github.com/user-attachments/assets/f0e76e79-18ca-4969-9396-2cfe2f44b632)

Once that is created, you need to get the signing secret by clicking the **"Reveal"**
![image](https://github.com/user-attachments/assets/0b5d311b-a99f-4dcd-98ab-f94da954c2e0)

And then copy & paste the signing secret in your **"Webhook Secret"** input.

Without the signing secret, Freeform will discard the event.


From the `404` error that your live site is displaying, it looks like the URL cannot be found by stripe. Is the webhook url used in Live publicly accessible?

**Comment by ndsurgenor** [comment]
Hi there, thanks for getting back to me!

Screenshot of live site settings from inside craft Freeform
<img width="1466" alt="Screenshot 2024-10-17 at 12 13 20" src="https://github.com/user-attachments/assets/73aa2ef2-4f2a-4e04-8ef2-0966afa111a1">

I note in your screenshot of the webhook that there are payment_intent triggers present which I do not have as the documentation mentioned otherwise; could this be the problem?
<img width="628" alt="Screenshot 2024-10-17 at 12 28 11" src="https://github.com/user-attachments/assets/801e1730-cf2b-42b7-b6f0-708e52b608e3">

**Comment by gustavs-gutmanis** [comment]
@ndsurgenor I noticed that the documentation does not reflect the latest changes in Freeform 5 that we introduced, and I'm very sorry for that. Could you add the payment intent events, and remove the other ones, please?

We will adjust the documentation and inline examples in Freeform. Sorry again.

**Comment by ndsurgenor** [comment]
No worries! I'll give it a go now and see how it goes :)

**Comment by gustavs-gutmanis** [comment]
Please let me know if that solves it. I have a feeling that it might not, since the `404` error indicates that Stripe cannot reach the endpoint, which is weird. It worked in my tests.
Are you able to access the webhook url via your browser?

**Comment by ndsurgenor** [comment]
Hi again. Updated those webhooks and used the $ stripe listen --forward-to https://your-local-url.test/freeform/payments/stripe/webhook --skip-verify --events=payment_intent.succeeded and testing setting on my dev site. In the terminal for dev site I can see the following:

```
> Ready! You are using Stripe API Version [2024-06-20]. Your webhook signing secret is [REDACTED] (^C to quit)
2024-10-17 13:36:56   --> payment_intent.succeeded [evt_3QAswVJkBfbOCys70VfTmq25]
2024-10-17 13:36:57  <--  [401] POST https://[REDACTED].ddev.site/freeform/payments/stripe/webhook [evt_3QAswVJkBfbOCys70VfTmq25]
```

Still getting same error of "Connection to remote host denied (please contact support at https://support.stripe.com/contact/)" on the dev webhook and 404 (not found) on the live one.

Putting the webhook URL into a browser gives a 404 page. Screenshot below from live site:
<img width="1703" alt="Screenshot 2024-10-17 at 13 43 50" src="https://github.com/user-attachments/assets/ebf33329-24c6-41da-90c8-ed9efd60f0fa">

**Comment by gustavs-gutmanis** [comment]
Hi @ndsurgenor,

### Dev URL

The `401` error happens when the signing secret is invalid.
Did you copy the signing secret from the terminal into the stripe integration **"Webhook Secret"** field?
Each time you run `stripe listen` a new secret is generated for the duration of the process, and you need to copy it into the stripe integration.

Also just to make sure - did you change the `https://your-local-url.test` to your local domain? I used `your-local-url.test` as an example, you have to change it to your local url.

### Public URL

As for the `404` error which you get on the public site - that is fine, since the `payment_intent` that is coming from a webhook will contain hashed data about the form, integration and submission used in the integration. If you just open the page in your browser, it won't have it and will throw a `404`, but that tells us that the endpoint is working as expected, instead of it being inaccessible. So your public site webhook should work fine, assuming that the signing secret and webhook destination URL's are configured properly.

**Comment by ndsurgenor** [comment]
> As for the 404 error which you get on the public site - that is fine, since the payment_intent that is coming from a webhook will contain hashed data about the form, integration and submission used in the integration. If you just open the page in your browser, it won't have it and will throw a 404, but that tells us that the endpoint is working as expected, instead of it being inaccessible. So your public site webhook should work fine, assuming that the signing secret and webhook destination URL's are configured properly.

Yes, when I try that URL in my browser it shows the custom 404 page for that site.

So you are saying not to worry about Stripe continuing to show  a 404 error for the payment_intent.succeeded event? The reason I have some concern is that when trying it before, Stripe automatically disabled the webhook due to so many failed attempts. Is there a way to tell Stripe to ignore the 404 and treat it as a 200?

**Comment by gustavs-gutmanis** [comment]
With the change in events that have to be selected for the webhook, 404 errors should no longer be happening.

**Comment by ndsurgenor** [comment]
So if Stripe is continuing to send and retry them as per the screenshot below this won't be an issue or I should try setting it up again to prevent this error?

<img width="1163" alt="Screenshot 2024-10-23 at 16 43 30" src="https://github.com/user-attachments/assets/6d81c4f5-a86d-4599-a9e7-fbbde7f5f531">

**Comment by gustavs-gutmanis** [comment]
No this should not be happening. Is this happening with the live version and an actual stripe submission?

**Comment by gustavs-gutmanis** [comment]
Could you please create a [support ticket](https://docs.solspace.com/support/) so we can see how we can help you out with this issue?

**Comment by seandelaney** [comment, marked as answer]
I was able to fully replicate and found the issue. PR submitted with fix.
